### PR TITLE
[WIP] oslo-config: 4.13.2 -> 5.1.0

### PR DESCRIPTION
**Pull request by makefu**
###### Motivation for this change

bump oslo-config to 5.1.0 and enable tests 💯 

For enabling tests i had to bump:
* reno: 2.3.2 -> 2.6.0 
* bandit: 0.16.0 -> 1.4.0
* openstackdocstheme: init at 1.17.0
* pyasn1-modules: 0.1.5 -> 0.2.1 
* ldappool: 1.0 -> 2.1.0 
* pythonPackages.ldap: 2.4.45 -> 3.0.0b2

Waiting for @GrahamcOfBorg :3
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [NA] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/bandit/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  };
+
+  preCheck = ''
+    export PATH=$PATH:$out/bin
+  '';
+
+  propagatedBuildInputs = [ pbr six pyyaml appdirs stevedore GitPython git ];
+  checkInputs = [ beautifulsoup4 oslosphinx testtools testscenarios
+                              testrepository fixtures mock ];
+  patchPhase = ''
```
Comment (FRidh) [inline pkgs/development/python-modules/bandit/default.nix]: `postPatch` is preferred so that one can still pass in `patches`.

**Review by FRidh — COMMENTED** [review]
Looks good to me.

**Comment by FRidh** [comment]
`oslo-config` bump is likely to break things though. That needs to be checked first.

**Comment by makefu** [comment]
i am currently on it, i will have to bump a number of libs to get it running. changing Headline to WIP

**Comment by makefu** [comment]
oh oh, it seems that bumping pyasn1 was not the wisest thing to do. it was for getting ldap to the latest version ... i will see how much stuff breaks with this.

**Comment by makefu** [comment]
can this PR be submitted to hydra for evaluation? i am afraid my pc is not hardcore enough to evaludate the PR in time. if it is not possible then i will probably have to revert and wait for pyasn1 be bumped by someone more adventurous than me.

**Comment by makefu** [comment]
for python3Packages.oslo-config i get
```
error: invalid command 'bdist_wheel'
```
however the package is building for `python2`. @FRidh you had similar issues before? i tried to put wheel in every `*Inputs` list without success.

**Comment by FRidh** [comment]
That typically happens when Python 2 and 3 packages get mixed during a build. Both packages have the interpreter in `propagatedBuildInputs`, and the interpreters each a setup hook that extend to the same `PYTHONPATH`. That causes trouble. In this case, certain Python 2 applications are added in the build when tests are enabled. I suggest for now disabling tests for Python 3 builds.

**Comment by FRidh** [comment]
When disabling tests, can you refer to the previous comment.

**Comment by makefu** [comment]
@FRidh i think the package is `reno` (which comes from `top-level/all-packages.nix`), the correct way would be to pull reno into pythonPackages again, right?

**Comment by FRidh** [comment]
@makefu if its just an application, and we're not interested in the modules, then no, then its not the correct way to pull it into `python-packages.nix`. But it is a pragmatic way to get your tests working ;-)

The correct way, but out of scope here, is to make sure Python applications don't show their `site-packages` to the hook.

**Comment by makefu** [comment]
@FRidh would it be okay to copy the derivation as pythonPackage (instead of pythonApplication) to `python-packages.nix` ? however i am not totally sure if it is possible to override the derivation in all-packages.nix to make the name look "correct" instead of `pythonXPackages.reno`

**Comment by makefu** [comment]
so many broken packages ... is it possible with `nox-review pr` to check for all packages which broke due to the pr?

**Comment by FRidh** [comment]
> so many broken packages ... is it possible with nox-review pr to check for all packages which broke due to the pr?

`nox-review` computes what needs to be rebuild and then builds it. You could use `niff` to just check what packages need to be rebuild. Its in Nixpkgs.
https://github.com/FRidh/niff

> a lot of issues due to argparse, not sure why this happens though.

A lot of (older) packages have an `install_requires` on `argparse`, even though `argparse` is part of stdlib in 2.7+. We typically patch that dependency out.

>  i would really fix the packages but moving them one-by-one to development/python-modules is quite some effort to go through

Absolutely. Many of the packages you touch haven't been used in a while. It may therefore also be wise to evaluate whether they could instead be dropped.


I would prefer to move this whole `openstack` stuff out into an overlay. Its a pain to maintain, and its presence puts a burden on others that are contributing as well.

**Comment by makefu** [comment]
Moving `openstack` out may be a good idea, this looks like a never ending rabbit hole of (circular) dependencies.

What i could think of is using `pypi2nix` and just freeze all required applications (with tests disabled) and refreeze on each version change.

**Comment by FRidh** [comment]
> What i could think of is using `pypi2nix` and just freeze all required applications (with tests disabled) and refreeze on each version change.

Won't help, because we need a coherent set that works for Python 2.7, 3.4, 3.5, 3.6 and preferably also PyPI. With `pypi2nix` you generate for one version.

But I thought that certain openstack libraries were already marked as broken, so I wouldn't imagine you having to upgrade them then.

**Comment by makefu** [comment]
Thing is, that the update resulted in more dependencies which resulted in new errors (and more update) and more dependencies. a never ending cycle.

Regarding the overlay, i think only a single python version would be enough to provide the functionality of openstack. This however would not solve the issue if people would actually want to use certain libraries in their project.

imho each module should come with a huge warning sign if it pulls dependencies from openstack ...

**Pull request closed without merging**